(global) Convert HTML entities to UTF-8 characters

npmccallum opened the pull request:
This helps XML parsers which don't handle HTML entities.

Files changed here:
data/tlg0057/tlg010/tlg0057.tlg010.perseus-eng1.xml
data/tlg0086/tlg009/tlg0086.tlg009.perseus-eng1.xml
data/tlg0086/tlg010/tlg0086.tlg010.perseus-eng1.xml
data/tlg0086/tlg025/tlg0086.tlg025.perseus-eng1.xml
data/tlg0086/tlg029/tlg0086.tlg029.perseus-eng1.xml
data/tlg0086/tlg034/tlg0086.tlg034.perseus-eng1.xml
data/tlg0086/tlg038/tlg0086.tlg038.perseus-eng1.xml
data/tlg0094/tlg001/tlg0094.tlg001.perseus-eng1.xml
data/tlg0094/tlg002/tlg0094.tlg002.perseus-eng1.xml
data/tlg0094/tlg003/tlg0094.tlg003.perseus-eng1.xml
data/tlg0099/tlg001/tlg0099.tlg001.perseus-eng2.xml
data/tlg0099/tlg001/tlg0099.tlg001.perseus-eng1.xml
data/tlg0526/tlg001/tlg0526.tlg001.perseus-eng1.xml
data/tlg0526/tlg002/tlg0526.tlg002.perseus-eng1.xml
data/tlg0526/tlg003/tlg0526.tlg003.perseus-eng1.xml

data/tlg0543/tlg001/tlg0543.tlg001.perseus-eng1.xml
data/tlg0555/tlg004/tlg0555.tlg004.perseus-grc1..xml
data/tlg1799/tlg001/tlg1799.tlg001.perseus-eng1.xml

lcerrato commented:
@npmccallum 
Thank you.
As we update works to CTS and EpiDoc compliance, entities are removed as part of that workflow. Any file ending in "grc1" is likely not updated to the current best practices. This includes work such as header completion, encoding, and quality control issues.

There cases where if entity conversions are done out of order this can cause problems (this is particularly true of the beta code file in this batch which is preserved for backup purposes and will not be fixed in the short term--I would recommend against any edits to this file).

Large multi-file PRs are not encouraged due to other ongoing conversion work and potential for conflicts. We generally work one corpus or author at a time. 

I can check and see if there are any potential conflicts here, but I would request that no changes to diod.hist09-10_gk.xml be made at this time. (https://github.com/PerseusDL/canonical-greekLit/issues/1405) I have added a PR with a comment on this file.

npmccallum commented:
@lcerrato Thanks for the summary!

I caught the issue with `diod.hist09-10_gk.xml` and wrote a script to convert it from beta code to unicode. This work does not update the file to best TEI practices. But it is at least forward progress. Would you like me to submit this?

This led me down a rabbit trail where I found a bunch of other beta code in other files. Is there a description somewhere of which files should be considered to meet quality standards?

Would you like me to split this into multiple PRs?

lcerrato commented:
@npmccallum 
Thank you, but the unconverted file would need to be merged with other data as it is a partial work so it is not a matter of conversion per se, but rather data that was not ported from the old collection to GitHub. 
Ideally, all of this work would be addressed within the usual conversion workflow.
If the file is going to cause issues, I can move to another repo for preservation (it was moved here for summer employees who would not otherwise have access).

lcerrato commented:
@npmccallum 
For works that have been transitioned to the latest encoding, the change logs will say if the file has been updated to CTS and EpiDoc compliance and the file will be visible in the Scaife Viewer. As a rule of thumb, anything that is grc1 is likely not updated to best practices, but there will be files that were converted early on that could use refinement (sometimes these have things like beta code in footnotes that was missed or other irregularities).

Any file with a plain text equivalent in the package release is complaint.
https://github.com/PerseusDL/canonical-greekLit/releases

Additionally, the HookTest output (found under Actions) shows passing (= updated) files.

lcerrato commented:
@npmccallum 
I am going to move this file out of here as it is just going to cause confusion. Perhaps you can either close and resubmit minus anything related to diod.hist09-10_gk.xml or I can close this out simply work on the global entity fix in another pass.
(Our testing regime seems broken at the moment so no changes as yet.)

Happy to continue the conversation and answer any questions or discuss other contributions.

npmccallum commented:
@lcerrato I have removed `data/tlg0060/tlg001/diod.hist09-10_gk.xml` from the commit. Please review.

lcerrato commented on data/tlg1799/tlg001/tlg1799.tlg001.perseus-eng1.xml:
This file has several sets of encoding issues which is to be expected in a mathematical text. I have logged ones that caught my eye, but mainly for further work and these changes are not required here.
NB that `<*>` was the mark our data entry teams used when they could not read the print or did not know how to encode something. Whoever converted this file was unaware of that and has rendered that as &lt;*&gt; which makes it harder to know what is going on.

lcerrato replied:
I did not complete a comprehensive review: too many changes.

lcerrato commented on data/tlg1799/tlg001/tlg1799.tlg001.perseus-eng1.xml:
```diff
@@ -807,51 +807,51 @@
- and attempted to clear up difficult and obscure passages, (2) that Ab&umacr;'l Waf&amacr; al-B&umacr;zj&amacr;n&itilde; (939-997) <quote>introduced unnecessary additions and left out many things of great importance and entirely necessary,</quote>
- inasmuch as he was too long in various places in Book VI. and too short in Book X. where he left out entirely the proofs of the <emph>apotomae</emph>, while he made an unsuccessful attempt to emend XII. 14, (3) that Ab&umacr; Ja`far al-Kh&amacr;zin (d. between 961 and 971) arranged the postulates excellently but <quote>disturbed the number and order of the propositions, reduced several propositi ...
+    <p>The <title>Fihrist</title> goes on to say that the work was next translated by Ishāq b. Hunain, and that this translation was improved by Thābit b. Qurra. This Abū Ya`qūb Is[hnull ]āq b. [Hnull ]unain b. Is[hnull ]āq al-`Ibādī (d. 910) was the son of the most famous of Arabic translators, Hunain b. Ishāq al-`Ibādī (809-873), a Christian and physician to the Caliph alMutawakkil (847-861). T ...
+) had seen in the Greek copy in his possession the proposition in Book I. which Thābit took credit for, and that this was confirmed by Na[znull ]īf, the physician, to whom Yuhannā had shown it. This proposition may have been wanting in Ishāq, and Thābit may have added it, but without claiming it as his own discovery<note n="3" place="unspecified" anchored="yes">Steinschneider, p. 88.</note>. As a ...
```
[znull ] = z + underdot

lcerrato replied:
also below

npmccallum replied:
Fixed.

lcerrato commented on data/tlg1799/tlg001/tlg1799.tlg001.perseus-eng1.xml:
```diff
@@ -868,77 +868,77 @@
-    <p>An advantage in the Arabic version is the omission of VII. Def. 10, although, as Iamblichus had it, it may have been deliberately omitted by the Arabic translator. Another advantage is the omission of the analyses and syntheses of XIII. 1-5; but again these may have been omitted purp&omacr;sely, as were evidently a number of porisms which are really necessary.</p>
-    <p>One or two remarks may be added about the Arabic versions as compared with one another. Al-[Hnull ]ajj&amacr;j's object seems to have been less to give a faithful reflection of the original than to write a useful and convenient mathematical text-book. One characteristic of it is the careful references to earlier propositions when their results are used. Such specific quotations of earlier  ...
+    <p>Heiberg considers that the Arabian translator had before him a MS. which was related to b, but diverged still further from the rest of our MSS. He does not think that there is evidence of the existence of a redaction of Books I.—X. similar to that of Books XI., XII. in b; for Klamroth observes that it is the Books on solid geometry (XI.—XIII.) which are more remarkable than the others for  ...
+    <p>An advantage in the Arabic version is the omission of VII. Def. 10, although, as Iamblichus had it, it may have been deliberately omitted by the Arabic translator. Another advantage is the omission of the analyses and syntheses of XIII. 1-5; but again these may have been omitted purpōsely, as were evidently a number of porisms which are really necessary.</p>
```
purpōsely seems like stray mark

npmccallum replied:
Fixed.

lcerrato reviewed (approved):
Several of these files have too many changes to comprehensively review in this format. 

I do see errant spaces around things like em dashes, which would typically be fixed at the time of conversion. I also see ligatures that do not require preservation. Middots all appear errant (but I could not check all works to confirm).

So a global change missed some of this refinement. It is good to have a "spot check" as it typically points to refinements we would not otherwise spot absent a careful read-through.

lcerrato commented:
@npmccallum 
A pull from master would be required due to updates to the testing regime made on the weekend.

lcerrato commented:
@npmccallum 
I appreciate all of the work that went into this. Please let me know if there are specific works that you would like prioritized for further updates.

npmccallum commented on data/tlg1799/tlg001/tlg1799.tlg001.perseus-eng1.xml:
```diff
@@ -503,25 +503,25 @@
     <p>(3) an addition to Pappus' remarks about the curvilineal angle which is equal to a right angle without being one<note n="8" place="unspecified" anchored="yes"><emph>ibid</emph>. p. 190, 9-23.</note>.</p>
     <p>The defence of Geminus against Carpus, who combated his view of theorems and problems, is also probably due to Proclus<note n="9" place="unspecified" anchored="yes"><emph>ibid</emph>. p. 243, 12-29.</note>, as well as an observation on I. 38 to the effect that I. 35-38 are really comprehended in VI. 1 as particular cases<note n="10" place="unspecified" anchored="yes"><emph>ibid</emph>. pp. ...
-    <p>Lastly, we can have no hesitation in attributing to Proclus himself (1) the criticism of Ptolemy's attempt to prove the parallel-postulate<note n="11" place="unspecified" anchored="yes"><emph>ibid</emph>. p. 368, 1-23.</note>, and (2) the other attempted proof given &iacute;n the same note<note n="12" place="unspecified" anchored="yes"><emph>ibid</emph>. pp. 371, 11-373, 2.</note> (on I. 2 ...
+    <p>Lastly, we can have no hesitation in attributing to Proclus himself (1) the criticism of Ptolemy's attempt to prove the parallel-postulate<note n="11" place="unspecified" anchored="yes"><emph>ibid</emph>. p. 368, 1-23.</note>, and (2) the other attempted proof given ín the same note<note n="12" place="unspecified" anchored="yes"><emph>ibid</emph>. pp. 371, 11-373, 2.</note> (on I. 29) and  ...
```
Fixed.

npmccallum commented on data/tlg1799/tlg001/tlg1799.tlg001.perseus-eng1.xml:
```diff
@@ -598,12 +598,12 @@
     <p>(The discovery of the ancient papyrus showing readings agreeing <pb n="53"/>with some, or with all, of the Theonine MSS. against P now makes it necessary to be very cautious in applying these criteria.)</p>
     <p>It is of course the last class (<foreign lang="greek">d</foreign>) of changes which we have to investigate in order to get a proper idea of Theon's recension.</p>
-    <p>Heiberg first observes, as regards these, that we shall find that Theon, in editing the <title>Elements</title>, altered hardly anything without some reason, often inadequate according to our ideas, but still some reason which seemed to him sufficient. Hence, in cases of very slight differences where both the Theonine MSS. and P have readings good and probable in themselves, Heiberg is not ...
+    <p>Heiberg first observes, as regards these, that we shall find that Theon, in editing the <title>Elements</title>, altered hardly anything without some reason, often inadequate according to our ideas, but still some reason which seemed to him sufficient. Hence, in cases of very slight differences where both the Theonine MSS. and P have readings good and probable in themselves, Heiberg is not ...
```
Fixed.

npmccallum commented on data/tlg1799/tlg001/tlg1799.tlg001.perseus-eng1.xml:
```diff
@@ -783,20 +783,20 @@
+    <p>V^{a} has, besides Schol. Vind., a number of scholia which also appear in other MSS., one in BFb, some others in P, and some in v (Codex Vat. 1038, 13th c.); these scholia were taken from a source in which many abbreviations were used, as they were often misunderstood by V^{a}. Other scholia in V^{a} which are not found in the older sources—some appearing in V^{a} alone—are also not origin ...
     <p>V^{b} seldom has scholia common with the other older sources; for the most part they either appear in V^{b} alone or only in the later sources as v or F^{2} (later scholia in F), some being original, others not. In Book X. V^{b} has three series of numerical examples, (1) with Greek numerals, (2) alternatives added later, also mostly with Greek numerals, (3) with Arabic numerals. The last  ...
-    <p>Of collections in other hands in V distinguished by Heiberg (see preface to Vol. v.), V^{1} has very few scholia which are found in other sources, the greater part being original; V^{2}, V^{3} are the work of the copyist himself; V^{4} are so in part only, and contain several scholia from Schol. Vat. and other sources. V^{3} and V^{4} are later than 13th &mdash;14th c., since they are not  ...
+    <p>Of collections in other hands in V distinguished by Heiberg (see preface to Vol. v.), V^{1} has very few scholia which are found in other sources, the greater part being original; V^{2}, V^{3} are the work of the copyist himself; V^{4} are so in part only, and contain several scholia from Schol. Vat. and other sources. V^{3} and V^{4} are later than 13th —14th c., since they are not found  ...
```
I think your request here was to remove the space before the mdash. But I'm not sure.

npmccallum commented on data/tlg1799/tlg001/tlg1799.tlg001.perseus-eng1.xml:
```diff
@@ -783,20 +783,20 @@
     <p>Heiberg has satisfied himself, by a closer study of b, that the scholia which he denotes by b, <foreign lang="greek">b</foreign> and b^{1} are by one hand; they are mostly to be found in other sources as well, though some are original. By the same hand (Theodorus Cabasilas, 15th c.) are also the scholia denoted by b^{2}, B^{2}, b^{3} and B^{3}. These scholia come in great part from Schol.  ...
     <p>As B^{3} (a late hand in B) contains several of the original scholia of b^{2}, B^{3} must have used b itself as his source, and, as all the scholia in B^{3} are in b, the latter is also the source of the scholia in B^{3} which are found in other MSS. B and b were therefore, in the 15th c., in the hands of the same person; this explains, too, the fact that b in a late hand has some scholia  ...
-    <p>An author's name is also attached to the scholia VI. No. 6 and X. No. 223, which are attributed to Maximus Planudes (end of 13th c.) along with scholia on I. 31, X. 14 and X. 18 found in 1 in a quite late hand and published on pp. 46, 47 of Heib&edot;rg's dissertation. These seem to have been taken from lectures of Planudes on the <title>Elements</title> by a pupil who used l as his copy.< ...
+    <p>An author's name is also attached to the scholia VI. No. 6 and X. No. 223, which are attributed to Maximus Planudes (end of 13th c.) along with scholia on I. 31, X. 14 and X. 18 found in 1 in a quite late hand and published on pp. 46, 47 of Heibėrg's dissertation. These seem to have been taken from lectures of Planudes on the <title>Elements</title> by a pupil who used l as his copy.</p>
```
Fixed.

npmccallum commented on data/tlg0526/tlg003/tlg0526.tlg003.perseus-eng1.xml:
```diff
@@ -755,7 +755,7 @@ to Hirom to be solved, and desired he would send others back for him to
 pay money to him that solved them. And when Hirom had agreed to the proposals,
 but was not able to solve the problems, he was obliged to pay a great deal
-of money, as a penalty for the same. As also they relate, that one&middot;Abdemon,
+of money, as a penalty for the same. As also they relate, that one·Abdemon,
```
Fixed.

npmccallum commented on data/tlg0526/tlg001/tlg0526.tlg001.perseus-eng1.xml:
```diff
@@ -648,7 +648,7 @@ them, as to the distance of 150 days to the 17th day of the 7th month,
 called by the Macedonians <foreign lang="xgreek">Dius</foreign>, but by the Hebrews <foreign lang="xhebrew">Marchesuan</foreign>:
 for so did they order their year in <placeName key="tgn,7016833">Egypt</placeName>. But <placeName key="tgn,2525806">Moses</placeName> appointed that
-&middot; <emph>Nisan</emph>, which is the same with Xanthicus, should be the
+· <emph>Nisan</emph>, which is the same with Xanthicus, should be the
```
Fixed.

npmccallum commented on data/tlg0526/tlg002/tlg0526.tlg002.perseus-eng1.xml:
```diff
@@ -1197,7 +1197,7 @@ But a few hours afterwards, I got up, and when I had dismissed the rest
 ordered my servant to get some wine ready. I also opened the letter so,
 that nobody could perceive it; and understanding thereby presently the
-purport&middot; of the writing, I sealed it up again, and appeared as if
+purport· of the writing, I sealed it up again, and appeared as if
```
Fixed.

npmccallum commented:
1. I rebased the changes on the latest master.
2. I split each file into a separate commit. This is a prelude to splitting into separate PRs, if you want.
2. I fixed most of the cleanups mentioned above.
3. I did not address the `<*>` or `superscript` issues as I'm not sure what precisely you want me to do with those, or the scope of the requested changes.

lcerrato commented:
@npmccallum 
I'll take a look. I think this is great for now.

npmccallum commented:
@lcerrato I filed this issue for `<*>` markers: https://github.com/PerseusDL/canonical-greekLit/issues/1529

npmccallum commented:
@lcerrato I filed this issue for the superscripts: https://github.com/PerseusDL/canonical-greekLit/issues/1530

npmccallum commented:
Closing in favor of separate PRs:

- #1531
- #1532
- #1533
- #1534
- #1535
- #1536
- #1538
- #1539
- #1540
- #1541
- #1542
- #1543
- #1544
- #1545
- #1546
- #1547
- #1548
- #1549

lcerrato commented:
I don't understand why this was closed when it was ready for merging?

npmccallum commented:
@lcerrato I found some other issues which are fixed in the split PRs. I think each one deserves individual review.

lcerrato commented:
@npmccallum Ok, thanks for the clarification. I was going to merge pending the earlier check.

Some/most of the files are too large to review in the GUI, so that probably won't be feasible on any scale in the short term, but almost all of these files will require additional review at time of conversion (they are mostly translations) so I was ready to merge here because I could not do a comprehensive check. 
(Work on tlg0555 corpus is underway at present)

npmccallum commented:
@lcerrato Helpful tip: review of these diffs is MUCH easier using `git diff --word-diff-regex=.`.
